codegen/src/types/composite_def.rs

```diff
@@ -27,16 +28,60 @@ pub struct CompositeDef {
 
 impl CompositeDef {
     /// Construct a definition which will generate code for a standalone `struct`.
+    ///
+    /// This is useful for generating structures from call and enum metadata variants;
+    /// and from all the runtime types of the metadata.
+    ///
+    /// # Note
+    ///
+    /// ## Alias Generation
+    ///
+    /// This method will generate alias types for each field of the struct when the
+    /// `generate_alias` parameter is true and the structure has any fields.
+    ///
+    /// The runtime types that are placed under `pub mod runtime_types` may contain
+    /// generics that are not resolved yet. Therefore, this field should be disabled
+    /// for runtime types.
+    ///
+    /// ### Example with alias generation
+    ///
+    /// This is extracted from the `system::calls` module.
+    ///
+    /// ```ignore
+    /// pub struct RemarkWithEvent {
+    ///     // Note that this type is aliased.
+    ///     pub remark: remark_with_event::Remark,
+    /// }
+    ///
+    /// pub mod remark_with_event {
+    ///     use super::runtime_types;
+    ///     pub type Remark = ::std::vec::Vec<::core::primitive::u8>;
+    /// }
+    /// ```
+    ///
+    /// ### Example without alias generation
+    ///
+    /// This is extracted from `runtime_types::pallet_referenda` module.
+    ///
+    /// ```ignore
+    /// pub struct DecidingStatus<_0> {
+    ///     pub since: _0,
+    ///     pub confirming: ::core::option::Option<_0>,
+    /// }
+    /// ```
     #[allow(clippy::too_many_arguments)]
     pub fn struct_def(
         ty: &Type<PortableForm>,
+        types_mod_ident: &syn::Ident,
         ident: &str,
+        variant_name: &str,
```

Comment:
I find it a little bit confusing to add a `variant_name` field here, which will only have a special meaning for call types etc... I suppose for most types this will be the same as the ident? 
Also I think we should maybe introduce a new struct for the type alias functionality instead of cramping it all into a CompositeDef. 

But maybe I don't understand enough why these decisions are made.

Reply:
Is it a bit scary to me that this already long list of arguments has grown by 3. It may be trying to do too many things now.

I wonder how easy it would be to separate the logic to generate the structs with aliases that we need in the "subxt codegen" side, and try to minimise how much we touch this `types` folder (which will be going away soon when `scale-typegen` is used instead)?

Reply:
Yep that makes sense!

I've split the implementation to rely on `TypeAliases` for generating the type alias module. 
On the `CompositeDef`, we still need a bit of information wrts whatever to generate the structure from scratch (needed for generating all types from the registry); or to generate the structure with type aliases (calls / events).

On that note, we'd probably need to think about integration with scale-typegen wrts:
- the current `CompositeDefFields` has a `to_type_aliases_tokens` which generates aliases from the fields of a structure 
- `to_struct_field_tokens` is adjusted to conditionally use the aliases or the types directly

@tadeohepperle let me know if this is right